daemon/emer-daemon.c

```diff
@@ -1123,6 +1142,379 @@ set_max_bytes_buffered (EmerDaemon *self,
   priv->max_bytes_buffered = max_bytes_buffered;
 }
 
+static void schedule_next_midnight_tick (EmerDaemon *self);
+static void schedule_next_month_tick (EmerDaemon *self);
+
+static GVariant *
+aggregate_timer_to_variant (AggregateTimerData *timer_data,
+                            guint32             previous_counter)
+{
+  gboolean complete = timer_data->end_monotonic_us != -1;
+  guint32 counter;
+  gint64 end_time;
+
+  counter = previous_counter;
+
+  end_time = complete ? timer_data->end_monotonic_us : g_get_monotonic_time ();
+  counter += (guint32) end_time - timer_data->start_monotonic_us;
+
+  return g_variant_new ("(vvvuubmv)",
+                        timer_data->event_id,
+                        timer_data->aggregate_key,
+                        timer_data->cache_key,
+                        timer_data->unix_user_id,
+                        counter,
+                        complete,
+                        timer_data->payload);
+}
+
+static gchar *
+get_tally_path_from_aggregate_timer (AggregateTimerData *timer_data)
+{
+
+  EmerDaemonPrivate *priv =
+    emer_daemon_get_instance_private (timer_data->daemon);
+  g_autofree gchar *aggregate_key_str = NULL;
+  g_autofree gchar *event_id_str = NULL;
+  g_autofree gchar *date_string = NULL;
+  g_autofree gchar *filename = NULL;
+
+  date_string = g_date_time_format (priv->current_aggregate_tally_date,
+                                    "%F");
+
+  aggregate_key_str = g_variant_print (timer_data->aggregate_key, FALSE);
+  event_id_str = g_variant_print (timer_data->event_id, FALSE);
+  filename = g_strdup_printf ("%s-%u-%s",
+                              aggregate_key_str,
+                              timer_data->unix_user_id,
+                              event_id_str);
+
+  return g_build_filename (priv->persistent_cache_directory,
+                           "aggregate-timers",
+                           date_string,
+                           filename,
+                           NULL);
```

Comment:
It's probably fine to have a file per aggregate key per user per event per day. I'd think some sort of indexed file/database would be nicer to use, particularly if there's a single reader and writer. I don't have any great ideas there, though.